channels/rdpgfx/server/rdpgfx_main.c

```diff
@@ -0,0 +1,1239 @@
+	{
+		if (BytesReturned == sizeof(HANDLE))
+			CopyMemory(&ChannelEvent, buffer, sizeof(HANDLE));
+
+		WTSFreeMemory(buffer);
+	}
+	else
+	{
+		WLog_ERR(TAG, "WTSVirtualChannelQuery failed");
+		error = ERROR_INTERNAL_ERROR;
+		goto out;
+	}
+
+	nCount = 0;
+	events[nCount++] = priv->stopEvent;
+	events[nCount++] = ChannelEvent;
+
+	/* Wait for the client to confirm that the dynamic channel is ready */
+	while (1)
+	{
+		if ((status = WaitForMultipleObjects(nCount, events, FALSE, 100)) == WAIT_OBJECT_0)
+			goto out;
+
+		if (status == WAIT_FAILED)
+		{
+			error = GetLastError();
+			WLog_ERR(TAG, "WaitForMultipleObjects failed with error %lu", error);
+			goto out;
+		}
+
+		if (WTSVirtualChannelQuery(priv->rdpgfx_channel, WTSVirtualChannelReady, &buffer, &BytesReturned) == FALSE)
+		{
+			WLog_ERR(TAG, "WTSVirtualChannelQuery failed");
+			error = ERROR_INTERNAL_ERROR;
+			goto out;
+		}
+
+		ready = *((BOOL*) buffer);
+
+		WTSFreeMemory(buffer);
+
+		if (ready)
+			break;
+	}
+
+	/* Create shared stream */
+	s = Stream_New(NULL, 4096);
+	if (!s)
+	{
+		WLog_ERR(TAG, "Stream_New failed!");
+		error = CHANNEL_RC_NO_MEMORY;
+		goto out;
+	}
+
+	/* Main virtual channel loop. RDPGFX do not need version negotiation */
+	while (ready)
+	{
+		status = WaitForMultipleObjects(nCount, events, FALSE, INFINITE);
+
+		if (status == WAIT_FAILED)
+		{
+			error = GetLastError();
+			WLog_ERR(TAG, "WaitForMultipleObjects failed with error %lu", error);
+			goto out;
+		}
+
+		/* Stop Event */ 
+		if (status == WAIT_OBJECT_0)
+			break;
+
+		/* Channel Event */
+		Stream_SetPosition(s, 0);
+
+		if (!WTSVirtualChannelRead(priv->rdpgfx_channel, 0, NULL, 0, &BytesReturned))
+		{
+			WLog_ERR(TAG, "WTSVirtualChannelRead failed!");
+			error = ERROR_INTERNAL_ERROR;
+			break;
+		}
+
+		if (BytesReturned < 1)
+			continue;
+
+		if (!Stream_EnsureRemainingCapacity(s, BytesReturned))
+		{
+			WLog_ERR(TAG, "Stream_EnsureRemainingCapacity failed!");
+			error = CHANNEL_RC_NO_MEMORY;
+			break;
+		}
+
+		if (WTSVirtualChannelRead(priv->rdpgfx_channel, 0, (PCHAR) Stream_Buffer(s),
+			Stream_Capacity(s), &BytesReturned) == FALSE)
+		{
+			WLog_ERR(TAG, "WTSVirtualChannelRead failed!");
+			error = ERROR_INTERNAL_ERROR;
+			break;
+		}
+
+		Stream_SetLength(s, BytesReturned);
+		Stream_SetPosition(s, 0);
+		if ((error = rdpgfx_server_receive_pdu(context, s)))
+		{
+			WLog_ERR(TAG, "rdpgfx_server_receive_pdu failed with error %lu!", error);
+			break;
+		}
+		Stream_SetPosition(s, 0);
+	}
+
+	Stream_Free(s, TRUE);
+out:
+	WTSVirtualChannelClose(priv->rdpgfx_channel);
+	priv->rdpgfx_channel = NULL;
+	if (error && context->rdpcontext)
+		setChannelError(context->rdpcontext, error, "rdpgfx_server_thread_func reported an error");
+
+	ExitThread((DWORD)error);
+	return NULL;
+}
+
+static BOOL rdpgfx_server_open(RdpgfxServerContext* context)
+{
+	RdpgfxServerPrivate* priv = (RdpgfxServerPrivate *) context->priv;
+
+	if (!priv->thread)
```

Comment:
error cases must be looked at because you leak some priv members depending of where it fails.